### PR TITLE
Add support for UUIDs and various ID generation strategies

### PR DESCRIPTION
**Pull request by dunglas**
Allow more advanced ID generation strategies:

* auto (autoincrement), by default
* server generated UUID
* client submitted UUID
* custom ID

Supports both Doctrine ORM and MongoDB.

Default: traditional autoincrement

UUID generated by the DBMS:

```yaml
id:
  generationStrategy: uuid
```

UUID submitted by the client:

```yaml
id:
  generationStrategy: uuid
  writable: true
```

ID (any string) submitted by the client:

```yaml
id:
  generationStrategy: none
```
Don't generate the ID field (ex: DTO):

```yaml
id:
  generate: false
```

### CODE REVIEW

File: src/TypesGenerator.php [hunk 1]
```diff
@@ -293,28 +293,55 @@ public function generate(array $config): void
...
             foreach ($classes as &$class) {
                 if ($class['hasChild'] || $class['isEnum'] || $class['embeddable']) {
                     continue;
                 }
 
+                switch ($config['id']['generationStrategy']) {
+                    case 'auto':
+                        $range = 'Integer';
+                        $typeHint = 'int';
+                        $writable = false;
```
Comment (Simperfit) [inline src/TypesGenerator.php]: can't `$writable` && `$nullable` be set by default ?

Comment (dunglas) [reply src/TypesGenerator.php]: If the strategy is auto, you cannot set the ID, by definition.

Comment (Simperfit) [reply src/TypesGenerator.php]: What I meant is that both strategy auto and mongoid share the same values

Comment (dunglas) [reply src/TypesGenerator.php]: Good catch, I'll merge both sections.

Comment (dunglas) [reply src/TypesGenerator.php]: No we can't merge both sections. MongoID is a string, auto an integer. I don't get what you mind.

Comment (Simperfit) [reply src/TypesGenerator.php]:
```
$range = 'Text';
                $typeHint = 'string';
                $writable = false;
                $nullable = true;
                
                switch ($config['id']['generationStrategy']) {
                    case 'auto':
                        $range = 'Integer';
                        $typeHint = 'int';
                        break;
                    case 'uuid':
                        $writable = $config['id']['writable'];
                        $nullable = !$writable;
                        break;
                    default:
                        $writable = true;
                        $nullable = !$writable;
                        break;
```

something like this

Comment (dunglas) [reply src/TypesGenerator.php]: Ok got it, I though about that but I find this less explicit and it introduces var reassignments. I prefer to keep it as is.

Comment (Simperfit) [reply src/TypesGenerator.php]: Yup, not a problem ;)

**Pull request merged**